accounts-db/src/accounts_db.rs

```diff
@@ -4042,14 +4054,16 @@ impl AccountsDb {
 
                         // If we are marking something dead, and the only remaining alive account is zero lamport, then make that zero lamport slot ready to be cleaned.
                         // If that slot happens to only contain zero lamport accounts, the whole slot will go away
-                        if slot_list.len() == 1 // should we also check for ref counts here?
-                            && slot_list
-                                .iter()
-                                .all(|(_slot, acct_info)| acct_info.is_zero_lamport())
-                        {
-                            adding_slots_to_clean += 1;
-                            self.accounts_index
-                                .add_uncleaned_roots(slot_list.iter().map(|(slot, _)| *slot));
+                        if slot_list.len() == 1 {
+                            // should we also check for ref counts here?
+                            if let Some((slot_alive, acct_info)) = slot_list.first() {
+                                if acct_info.is_zero_lamport() && !acct_info.is_cached() {
+                                    self.zero_lamport_single_ref_found(
```

Comment:
this hsould probably ALSO run when we are doing clean and we find a zero lamport account that has a single ref. This is a reason why the hashset may be necessary so we don't double-count it.

Reply:
@HaoranYi